### PR TITLE
[Quantization] add humming mxfp4 moe backend

### PR DESCRIPTION
**Pull request by jinzhen-lin**
This PR add humming mxfp4 moe backend.

Humming project: https://github.com/inclusionAI/humming/

In https://github.com/vllm-project/vllm/pull/34556 , we add an initial integration of humming. Now I am working to integrate humming backends to dense/moe kernel oracles.

This backend supports running DeepSeek-V4.

Note for users: You must pass the `--moe-backend humming` flag to use this backend, as Humming is not currently a mandatory dependency for vLLM. I am working on releasing the Humming PyPI package. Until then, you can install it using:

```
pip install git+https://github.com/inclusionAI/humming.git
```

## Benchmark

DeepSeek-V4-Flash + H20 x 4 

Service start command:

```
# Marlin W4A16 (Baseline) (Use command from vllm recipes)
vllm serve /home/admin/DeepSeek-V4-Flash/ \
  --served-model-name model \
  --trust-remote-code \
  --kv-cache-dtype fp8 \
  --block-size 256 \
  --enable-expert-parallel \
  --data-parallel-size 4 \
  --compilation-config '{"cudagraph_mode":"FULL_AND_PIECEWISE", "custom_ops":["all"]}' \
  --tokenizer-mode deepseek_v4 \
  --tool-call-parser deepseek_v4 \
  --enable-auto-tool-choice \
  --reasoning-parser deepseek_v4


# Humming W4A16
vllm serve /home/admin/DeepSeek-V4-Flash/ \
  --served-model-name model \
  --trust-remote-code \
  --kv-cache-dtype fp8 \
  --block-size 256 \
  --enable-expert-parallel \
  --data-parallel-size 4 \
  --compilation-config '{"cudagraph_mode":"FULL_AND_PIECEWISE", "custom_ops":["all"]}' \
  --tokenizer-mode deepseek_v4 \
  --tool-call-parser deepseek_v4 \
  --enable-auto-tool-choice \
  --reasoning-parser deepseek_v4 \
  --moe-backend humming


# Humming W4A8
VLLM_HUMMING_INPUT_QUANT_CONFIG='{"dtype": "float8e4m3"}' vllm serve /home/admin/DeepSeek-V4-Flash/ \
  --served-model-name model \
  --trust-remote-code \
  --kv-cache-dtype fp8 \
  --block-size 256 \
  --enable-expert-parallel \
  --data-parallel-size 4 \
  --compilation-config '{"cudagraph_mode":"FULL_AND_PIECEWISE", "custom_ops":["all"]}' \
  --tokenizer-mode deepseek_v4 \
  --tool-call-parser deepseek_v4 \
  --enable-auto-tool-choice \
  --reasoning-parser deepseek_v4 \
  --moe-backend humming
```

Bench command:

```
# for prefill
vllm bench serve \
    --model model \
    --host 127.0.0.1 \
    --tokenizer /home/admin/DeepSeek-V4-Flash/ \
    --num-prompts 128 \
    --random-input-len 65536 \
    --random-output-len 1 \
    --trust-remote-code \
    --max-concurrency 32

# for decoding
vllm bench serve \
    --model model \
    --host 127.0.0.1 \
    --tokenizer /home/admin/DeepSeek-V4-Flash/ \
    --num-prompts 256 \
    --random-input-len 1 \
    --random-output-len 4096 \
    --trust-remote-code \
    --max-concurrency 64
```

Bench result (TPS):

| | Prefill | Decoding |
| - | - | - |
| Marlin W4A16 | 10532.12 | 1650.82  |
| Humming W4A16 | 15052.24 | 1727.73  |
| Humming W4A8 | 19404.33  | 1730.07 |

The performance gains are primarily driven by enhancements in the moe kernel and the moe sum kernel.

marlin moe w4a16 kernel

image

humming moe w4a16 kernel

image

marlin moe sum kernel (torch.sum)

image

humming moe sum kernel (introduced in https://github.com/vllm-project/vllm/pull/34556)

image

## Accuracy Test

```
lm_eval \
    --model local-chat-completions \
    --tasks gsm8k \
    --num_fewshot 2 \
    --batch_size auto \
    --model_args "model=model,base_url=http://localhost:8000/v1/chat/completions,num_concurrent=128"
```

Marlin W4A16 (main)

```
Requesting API: 100%|██████████| 1319/1319 [03:23<00:00,  6.47it/s]
2026-04-28:21:01:31 INFO     [loggers.evaluation_tracker:316] Output path not provided, skipping saving results aggregated
local-chat-completions ({'model': 'model', 'base_url': 'http://localhost:8000/v1/chat/completions', 'num_concurrent': 128}), gen_kwargs: ({}), limit: None, num_fewshot: 2, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     2|exact_match|↑  |0.9272|±  |0.0072|
|     |       |strict-match    |     2|exact_match|↑  |0.1486|±  |0.0098|
```

Humming W4A16 (PR)

```
Requesting API: 100%|██████████| 1319/1319 [02:20<00:00,  9.40it/s]
2026-04-28:21:07:06 INFO     [loggers.evaluation_tracker:316] Output path not provided, skipping saving results aggregated
local-chat-completions ({'model': 'model', 'base_url': 'http://localhost:8000/v1/chat/completions', 'num_concurrent': 128}), gen_kwargs: ({}), limit: None, num_fewshot: 2, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     2|exact_match|↑  |0.9227|±  |0.0074|
|     |       |strict-match    |     2|exact_match|↑  |0.1471|±  |0.0098|
```

Humming W4A8 (PR)

```
Requesting API: 100%|██████████| 1319/1319 [02:21<00:00,  9.33it/s]
2026-04-28:21:12:52 INFO     [loggers.evaluation_tracker:316] Output path not provided, skipping saving results aggregated
local-chat-completions ({'model': 'model', 'base_url': 'http://localhost:8000/v1/chat/completions', 'num_concurrent': 128}), gen_kwargs: ({}), limit: None, num_fewshot: 2, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     2|exact_match|↑  |0.9219|±  |0.0074|
|     |       |strict-match    |     2|exact_match|↑  |0.1433|±  |0.0097|
```

### CODE REVIEW

**Comment by huangzhilin-hzl** [comment]
@jinzhen-lin Nice job. Could you add the corresponding accuracy comparison too?

**Comment by jinzhen-lin** [comment]
Hi @huangzhilin-hzl , thank you for your interest. The accucry tests are done now.

**Review by mgoin — APPROVED** [review]
Excellent work! Validated locally on H100 DSV4 with the latest humming

File: vllm/model_executor/layers/fused_moe/fused_humming_moe.py [hunk 1]
```diff
@@ -60,49 +59,44 @@ class HummingExpertsBase(mk.FusedMoEExpertsModular):
     def __init__(
         self,
         layer: torch.nn.Module,
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/fused_humming_moe.py:61]: We should try to avoid passing the layer here if at all possible.  It contains the modular kernels.  If we ever construct the modular kernels at `__init__` time of the layer (which we are considering) then this will lead to all sorts of problems.

Comment (jinzhen-lin) [reply vllm/model_executor/layers/fused_moe/fused_humming_moe.py:61]: Since Humming supports a wide variety of quantization combinations, the corresponding weight combinations are also quite numerous. To reduce the complexity on the caller side, I prefer to use a layer-based approach. If directly passing the `FusedMoE` layer would cause issues, do you think it would be a good choice to directly extract all the required weights and reconstruct a temporary layer inside the modular kernels.

Comment (jinzhen-lin) [reply vllm/model_executor/layers/fused_moe/fused_humming_moe.py:61]: I don't quite understand what "construct the modular kernels at `__init__` time of the layer" means. Since the modular kernels currently require passing in a `FusedMoEQuantConfig`, and this config can only be fully defined after `process_weights_after_loading`, how are we supposed to construct the modular kernels at the `__init__` stage? Do you plan to pass these in as runtime variables?

Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/fused_humming_moe.py:61]: > I don't quite understand what "construct the modular kernels at `__init__` time of the layer" means. Since the modular kernels currently require passing in a `FusedMoEQuantConfig`, and this config can only be fully defined after `process_weights_after_loading`, how are we supposed to construct the modular kernels at the `__init__` stage? Do you plan to pass these in as runtime variables?

Even though the modular kernels require a `FusedMoEQuantConfig` at construction time, they don't really need much information from it (if any).  We've been discussing removing this as a requirement for construction so that modular kernels can be instantiated at the same time as the quant methods that own them.  This is to address other subtle order of initialization issues related to the `FusedMoE` layer, quant methods, SharedExperts, MoERunner, etc.

Comment (jinzhen-lin) [reply vllm/model_executor/layers/fused_moe/fused_humming_moe.py:61]: So, are you planning to pass model parameters or layers as arguments to the apply function? (Many quantization methods have additional parameters besides weight and scale.) I can do the relevant refactoring work for humming in advance.

Comment (jinzhen-lin) [reply vllm/model_executor/layers/fused_moe/fused_humming_moe.py:61]: So, are you planning to pass model parameters or layers as arguments to the apply function? (Many quantization methods have additional parameters besides weight and scale.) I can do the relevant refactoring work for humming in advance.

Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/fused_humming_moe.py:61]: No, the layer will still be passed as a runtime arg to apply.  It's only a problem when used as an argument to `__init__` any modular kernel objects.

Comment (jinzhen-lin) [reply vllm/model_executor/layers/fused_moe/fused_humming_moe.py:61]: Hi @bnellnm ,  since layer is not yet a parameter for methods like `apply`, `moe_problem_size`, or `workspace_shapes`, I can't remove it from the `__init__` arguments for now. Is there a timeline for the refactoring you mentioned?

Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/fused_humming_moe.py:61]: > Hi @bnellnm , since layer is not yet a parameter for methods like `apply`, `moe_problem_size`, or `workspace_shapes`, I can't remove it from the `__init__` arguments for now. Is there a timeline for the refactoring you mentioned?

It looks like a number of the values you need are members of `FusedMoEConfig` or `FusedMoEQuantConfig` which each MK has as attributes, e.g. hidden_size, number of experts, etc.

The layer is not going to be passed down to the MK apply function.  It is passed to the quant_method `apply` functions and it is the quant_method's responsibility to unpack any data needed from the layer and pass it along to the MK.

Comment (jinzhen-lin) [reply vllm/model_executor/layers/fused_moe/fused_humming_moe.py:61]: Currently, the parameters for the MK apply function are fixed. Is there a way to pass specific parameters to the kernel? For example, the Humming MoE requires a `layer` object, and similarly, the Marlin MoE requires variables like `w13_g_idx`.

Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/fused_humming_moe.py:61]: > Currently, the parameters for the MK apply function are fixed. Is there a way to pass specific parameters to the kernel? For example, the Humming MoE requires a `layer` object, and similarly, the Marlin MoE requires variables like `w13_g_idx`.

From looking at `fused_humming_moe.py`, I see the following attributes being accessed from the layer:
```
self.layer.activation,
self.layer.expert_map,
self.layer.global_num_experts
self.layer.hidden_size
self.layer.humming_metas["w13"].num_experts
self.layer.humming_metas["w13"].shape_n
self.layer.humming_metas["w13"].shape_k
self.layer.humming_metas["w13"].a_dtype
self.layer.humming_metas["w13"].c_dtype
self.layer.intermediate_size_per_partition
self.layer.num_experts
self.layer.param_dtype
```
These should all be available in `FusedMoEConfig`, `FusedMoEQuantConfig` or as an argument to `apply`.

I don't know what's in the humming module so I can't tell what extra parameters are being used. I'd suggest putting them in a new `dataclass`/`torch.nn.Module` and passing that as a constructor argument to `HummingExperts` classes rather than the `FusedMoE` layer.

Comment (jinzhen-lin) [reply vllm/model_executor/layers/fused_moe/fused_humming_moe.py:61]: I'll try to refactor this. Can I reach you on Slack afterwards ?

Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/fused_humming_moe.py:61]: > I'll try to refactor this. Can I reach you on Slack afterwards ?

Thanks.  I'm around on the vllm slack if you need some help.
